###################################################################
# This script processes data from
# /gsap/garage-protistvector/tstraub/ampseq/gtseq_v1/amp_frqs.rds
# generated by Tim Straub 30th Jan 2020 & accessed by Aimee on Feb
# 3rd 2020 via
# scp aimeet@login.broadinstitute.org:/gsap/garage-protistvector/tstraub/ampseq/gtseq_v1/amp_frqs.rds ~/Documents/
#
# Tim's notes (copied from slack): This file contains Senegal,
# Columbia, French Guiana, and Mali data. For some amplicons,
# there was no variation in some countries, so I put in Genotype.1
# as a value of 1. A handful of amplicons had no useable SNP data
# and they were excluded from this file
# (PF3D7_1008700, PF3D7_1408100_1, PfHRP2_1, PfHRP3_1, and PvDHFR [vivax, not falciparum]).
# To make it easier to separate out the CTSA amplicons (CSP, TRAP, SERA2, and AMA1)
# from the regular gt-seq panel, those have a “z” pre-pended to their name and appear
# on the bottom of the data set (e.g. zTRAP). The genotypes have been
# ordered based on decreasing allele frequency, just in case that makes a difference.
#
# ************************ IMPORTANT ***************************
# In the following we refer to Tim's Genotype.1 as Allele.1 etc.
# As mentioned above, alleles per amplicon are ordered by their frequencies
# (this is a requirement of the HMM of [1]). As such, the names of
# the alleles per amplicon (Allele.1, Allele.2, etc.) likely correspond
# to different microhaplotype SNP sequences in different countries.
# e.g. the most common allele in Sengal may differ to that in
# Colombia. This is not a problem since data on each country
# should be simulated and analysed separately using the HMM of [1].
# It would be a problem if we estimated relatedness between parasite
# from different populations, however, e.g. using hmmIBD.
# **************************************************************
###################################################################
rm(list = ls())
require(plyr) # for dlply

# Process the example raw input data in prep. for panel data simulation ============

# Set working directory to this file location and load example raw input data
# The example raw input data include marker positions and allele frequency estimates.
# The markers are microhaplotypes that are typed using amplicons, hence "amp_freq".
# Note that microhaplotypes are not point mutations: they are regions of the genome
# that are highly diverse. To simulate microhaplotype data under the HMM model [1]
# each marker region is reduced to a single point: its middle position.
amp_frqs = readRDS('amp_frqs.rds')

# Check that there are no factors
any(apply(amp_frqs, 2, is.factor))

# Rename "Genotype.1" by "Allele.1" etc.
colnames(amp_frqs) <- gsub("Genotype", "Allele", colnames(amp_frqs))

# Separate amplicon data (marker regions) from country specific data (allele frequencies)
markers = amp_frqs[!duplicated(amp_frqs$Amplicon_name), c('Amplicon_name','Chr','Start','Stop')]
rownames(markers) = markers$Amplicon_name

# Add the middle position of each marker region
markers$pos = apply(markers[,c('Start','Stop')], 1, function(x)median(as.numeric(x)))

# Add numeric chromosome
markers$chrom = as.numeric(gsub('_v3', '', gsub('Pf3D7_', '', markers$Chr)))

# Manually check all chromosomes are represented: yes
all(sort(unique(markers$chrom)) == 1:14)

# Sort markers by order of increasing chromosome and position
reordered_data_list = list()
for(chr in sort(unique(markers$chrom))){
  x = markers$pos[chr == markers$chrom] # Extract positions per chromosome
  if(!all(x == cummax(x))){ # If not all monotonically increasing...
    inds = sort.int(x, index.return = T)$ix # Sort positions
    reordered_data_list[[chr]] = markers[chr == markers$chrom, ][inds, ]
  }
}

# Check re-ordered data: returns TRUE if ordered correctly
markers = do.call(rbind, reordered_data_list)
all(sapply(unique(markers$chrom), function(chr){
  x = markers$pos[chr == markers$chrom]
  all(x == cummax(x))
}))

# Compute distances between middle positions
markers$dt <- c(diff(markers$pos), Inf)
pos_change_chrom <- 1 + which(diff(markers$chrom) != 0) # find places where chromosome changes
markers$dt[pos_change_chrom-1] <- Inf

# Extract allele names
Alleles = names(amp_frqs)[grepl('Allele', names(amp_frqs))]

# Separate frequencies by country and check alleles are ordered by frequency per amplicon
ordered_markers = rownames(markers)
frequencies = plyr::dlply(amp_frqs, 'Country', function(x){
  rownames(x) = x$Amplicon_name
  y = x[ordered_markers,] # Order amplicons according to chrom and pos
  # Check order of alleles per amplicon and re-order if necessary
  if(!all(apply(y[, Alleles], 1, function(z){all(z == cummin(z))}))){
    y <- t(apply(y[, Alleles], 1, function(f) sort(as.numeric(f), decreasing = T))) # Order frequencies as numeric
  }
  return(y)
})

# Save example data
save(markers, frequencies, file = '../data/panel_markers_and_frequencies.RData')

